core/src/main/scala/sttp/tapir/server/interpreter/EncodeOutputs.scala

```diff
@@ -59,19 +61,41 @@ class EncodeOutputs[B, S](rawToResponseBody: ToResponseBody[B, S]) {
         )
       case EndpointOutput.OneOf(mappings, _) =>
         val enc = encoded[Any]
-        val mapping = mappings
-          .find(mapping => mapping.appliesTo(enc))
-          .getOrElse(throw new IllegalArgumentException(s"No status code mapping for value: $enc, in output: $output"))
-        apply(mapping.output, ParamsAsAny(enc), mapping.statusCode.map(ov.withStatusCode).getOrElse(ov))
+
+        val bodyMappings: Map[MediaType, StatusMapping[_]] = mappings
+          .filter(_.appliesTo(enc))
+          .flatMap(sm =>
+            sm.output.traverseOutputs {
+              case EndpointIO.Body(bodyType, codec, _) =>
+                Vector[(MediaType, StatusMapping[_])](
+                  charset(bodyType).map(ch => codec.format.mediaType.charset(ch.name())).getOrElse(codec.format.mediaType) -> sm
+                )
+              case EndpointIO.StreamBodyWrapper(StreamBodyIO(_, codec, _, charset)) =>
+                Vector[(MediaType, StatusMapping[_])](
+                  charset.map(ch => codec.format.mediaType.charset(ch.name())).getOrElse(codec.format.mediaType) -> sm
+                )
+              case EndpointIO.Empty(codec, _) => Vector[(MediaType, StatusMapping[_])](codec.format.mediaType -> sm)
+            }
+          )
+          .toMap
+
+        if (bodyMappings.nonEmpty) {
+          val ranges = Accepts.unsafeParse(requestHeaders)
+          val mediaTypes = bodyMappings.keys.toSeq.asInstanceOf[scala.collection.immutable.Seq[MediaType]]
+          MediaType
+            .bestMatch(mediaTypes, ranges)
+            .flatMap(bodyMappings.get)
+            .map(sm => apply(sm.output, ParamsAsAny(enc), sm.statusCode.map(ov.withStatusCode).getOrElse(ov)))
+            .getOrElse(ov.withStatusCode(StatusCode.NotAcceptable))
```

Comment:
I'm not sure if we can do this here. When we get to `EncodeOutputs`, the server logic has already run - can we return a 4xx response at this point? Shouldn't we either:
* perform the validation upfront, when decoding the requests - although we don't know what's the status code going to be, so we can only say if all or none match
* when no mapping is found, returning any response instead of "not acceptable"

Would be good to see how other libraries/frameworks behave in this case